apps/emqx_retainer/src/emqx_retainer.erl

```diff
@@ -324,24 +336,27 @@ read_message(Context, Topic) ->
     BackendState = backend_state(Context),
     Mod:read_message(BackendState, Topic).
 
--spec page_read(context(), emqx_maybe:t(topic()), non_neg_integer(), non_neg_integer()) ->
+-spec page_read(context(), emqx_maybe:t(topic()), deadline(), non_neg_integer(), non_neg_integer()) ->
     {ok, has_next(), list(message())}.
-page_read(Context, Topic, Page, Limit) ->
+page_read(Context, Topic, Deadline, Page, Limit) ->
     Mod = backend_module(Context),
     BackendState = backend_state(Context),
-    Mod:page_read(BackendState, Topic, Page, Limit).
+    Mod:page_read(BackendState, Topic, Deadline, Page, Limit).
 
 -spec count(context()) -> non_neg_integer().
 count(Context) ->
     Mod = backend_module(Context),
     BackendState = backend_state(Context),
     Mod:size(BackendState).
 
--spec clear_expired(context()) -> ok.
-clear_expired(Context) ->
-    Mod = backend_module(Context),
-    BackendState = backend_state(Context),
-    ok = Mod:clear_expired(BackendState).
+-spec start_clear_expired(context()) -> ok.
+start_clear_expired(Context) ->
+    Opts = #{
+        deadline => erlang:system_time(millisecond),
+        limit => emqx_conf:get([retainer, msg_clear_limit], all)
+    },
+    _Result = emqx_retainer_sup:start_gc(Context, Opts),
```

Comment:
Should we monitor the process and not start the next iteration until the previous one is finished?
E.g. to avoid piling up the cleaners.

Reply:
Supervisor starts it under the same child id each time, so in this case `_Result` should be `{error, {already_started, ...}}` (appreciate a double-check of this claim).